src/hd/hd.c

```diff
@@ -6122,17 +6122,39 @@ char* get_sysfs_attr(const char* bus, const char* device, const char* attr)
 /*
  * must be able to read more than one line
  */
-char *get_sysfs_attr_by_path(const char* path, const char* attr)
+char *get_sysfs_attr_by_path(const char *path, const char *attr)
 {
-  static char buf[1024];
-  int i, fd;
+  return get_sysfs_attr_by_path2(path, attr, NULL);
+}
+
+
+/*
+ * binary data version; return data length, too
+ */
+char *get_sysfs_attr_by_path2(const char *path, const char *attr, unsigned *len)
+{
+  static char *buf = NULL;
+  char * ptr;
+  int i, fd, max;
+
+  if(len) *len = 0;
+
+  // init static buffer on first run
+  if(!buf) buf = new_mem(MAX_ATTR_SIZE + 1);
+
+  if(!buf) return NULL;
 
   sprintf(buf, "%s/%s", path, attr);
   fd = open(buf, O_RDONLY);
   if(fd >= 0) {
-    i = read(fd, buf, sizeof buf - 1);
+    max = MAX_ATTR_SIZE;
+    ptr = buf;
+    while((i = read(fd, ptr, max)) > 0) max -= i, ptr += i;
```

Comment:
This `i` should really be something like bytes_read; this is misleading here since commonly a variable named `i` is used as a loop counter.

But since this is a backport, better leave it as it is to keep the diff to the original version short.